src/theory/sets/infer_proof_cons.cpp

```diff
@@ -0,0 +1,331 @@
+namespace theory {
+namespace sets {
+
+InferProofCons::InferProofCons(Env& env, TheorySetsRewriter* tsr)
+    : EnvObj(env), d_tsr(tsr), d_imap(userContext()), d_expMap(userContext())
+{
+  d_false = nodeManager()->mkConst(false);
+  d_tid = builtin::BuiltinProofRuleChecker::mkTheoryIdNode(THEORY_SETS);
+}
+
+void InferProofCons::notifyFact(const Node& conc,
+                                const Node& exp,
+                                InferenceId id)
+{
+  Assert(conc.getKind() != Kind::AND && conc.getKind() != Kind::IMPLIES);
+  d_imap[conc] = id;
+  d_expMap[conc] = exp;
+}
+
+void InferProofCons::notifyConflict(const Node& conf, InferenceId id)
+{
+  d_imap[conf.notNode()] = id;
+}
+
+void InferProofCons::notifyLemma(const Node& lem, InferenceId id)
+{
+  d_imap[lem] = id;
+}
+
+std::shared_ptr<ProofNode> InferProofCons::getProofFor(Node fact)
+{
+  NodeInferenceMap::iterator it = d_imap.find(fact);
+  Assert(it != d_imap.end());
+  InferenceId id = it->second;
+
+  // temporary proof
+  CDProof cdp(d_env);
+  std::vector<Node> assumps;
+  Node conc = fact;
+  // First split into conclusion and assumptions.
+  if (fact.getKind() == Kind::IMPLIES || fact.getKind() == Kind::NOT)
+  {
+    if (fact[0].getKind() == Kind::AND)
+    {
+      assumps.insert(assumps.begin(), fact[0].begin(), fact[0].end());
+    }
+    else
+    {
+      assumps.push_back(fact[0]);
+    }
+    if (fact.getKind() == Kind::IMPLIES)
+    {
+      conc = fact[1];
+    }
+    else
+    {
+      conc = d_false;
+    }
+    cdp.addStep(fact, ProofRule::SCOPE, {conc}, {assumps});
+  }
+  else
+  {
+    NodeExpMap::iterator itex = d_expMap.find(fact);
+    if (itex != d_expMap.end())
+    {
+      Node exp = itex->second;
+      if (exp.getKind() == Kind::AND)
+      {
+        assumps.insert(assumps.end(), exp.begin(), exp.end());
+      }
+      else
+      {
+        assumps.push_back(exp);
+      }
+    }
+  }
+  // Try to convert.
+  if (!convert(cdp, id, assumps, conc))
+  {
+    cdp.addTrustedStep(conc, TrustId::THEORY_INFERENCE, assumps, {d_tid});
+  }
+  return cdp.getProofFor(fact);
+}
+
+bool InferProofCons::convert(CDProof& cdp,
+                             InferenceId id,
+                             const std::vector<Node>& assumps,
+                             const Node& conc)
+{
+  // these are handled manually
+  Assert(id != InferenceId::SETS_PROXY
+         && id != InferenceId::SETS_PROXY_SINGLETON);
+  Trace("sets-ipc") << "InferProofCons::convert " << id << std::endl;
+  Trace("sets-ipc") << "- assumptions: " << assumps << std::endl;
+  Trace("sets-ipc") << "- conclusion:  " << conc << std::endl;
+  bool success = false;
+  TheoryProofStepBuffer psb(cdp.getManager()->getChecker(), true);
+  switch (id)
+  {
+    case InferenceId::SETS_DOWN_CLOSURE:
+    case InferenceId::SETS_MEM_EQ:
+    case InferenceId::SETS_MEM_EQ_CONFLICT:
+    {
+      Assert(assumps.size() >= 1);
+      // (and (set.member x S) (= S (op T1 T2))) =>
```

Comment:
Is it an invariant that the first assumption is always a membership constraint, and the remaining assumptions are equalities between set terms? If so, please add assertions here, since this can easily breaks if the order of assumptions change.